File: data-structures/4_hash_map/1_hashmapbasic/hashmapbasic.ts
class HashMapBasic {
  bucket: Array<any>;
  size: number = 100;

  constructor() {
    this.bucket = new Array(this.size);
  }

  hashKey(key: string) {
    let hash = 0;
    for (let i of key) {
      hash += i.charCodeAt(0);
    }
    return hash % this.size;
  }

  set(key: string, value: any) {
    const hash = this.hashKey(key);
    this.bucket[hash] = value;
  }

  get(key: string) {
    const hash = this.hashKey(key);
    return this.bucket[hash];
  }

  delete(key: string) {
    const hash = this.hashKey(key);
    delete this.bucket[hash];
  }

  containsKey(key: string) {
    const hash = this.hashKey(key);
    return this.bucket[hash] !== undefined;
  }
}

const hashmap = new HashMapBasic();
hashmap.set("name", "Hyderabad");
hashmap.set("age", 2000);

console.log(hashmap.get("name"));
console.log(hashmap.containsKey("name"));

hashmap.delete("name");

console.log(hashmap.get("name"));
console.log(hashmap.containsKey("name"));

File: data-structures/4_hash_map/1_hashmapbasic/hashmapbasic.py
class HashMapBasic:
    def __init__(self):
        self.bucketSize = 100
        self.bucket = [None] * self.bucketSize

    # Generate hash
    def getHash(self,key):
        asciiTot = 0
        for alpha in key:
            asciiTot+= ord(alpha)
        return asciiTot % len(self.bucket)
    
    def __setitem__(self, key, value):
        hash = self.getHash(key)
        self.bucket[hash] = value

    def __getitem__(self, key):
        hash = self.getHash(key)
        return self.bucket[hash]
    
    def delete(self, key):
        # We aren't using del or pop() as they remove the element and bucket size decreases
        hash = self.getHash(key)
        self.bucket[hash] = None 
    
    def contains(self, key):
        hash = self.getHash(key)
        return self.bucket[hash] is not None

hashMap = HashMapBasic()
hashMap["name"] = "chandrahas"
hashMap["age"] = 34
print(hashMap["name"], hashMap["age"])
hashMap.delete("name")
print(hashMap["name"])
print(hashMap.contains("name"), hashMap.contains("age"))
